compiler/modules/nback/build.bal

```diff
@@ -916,41 +916,167 @@ function buildBitwiseBinary(llvm:Builder builder, Scaffold scaffold, bir:IntBitw
 }
 
 function buildCompare(llvm:Builder builder, Scaffold scaffold, bir:CompareInsn insn) returns BuildError? {
-    match insn.orderType {
+    var [lhsRepr, lhsValue] = check buildReprValue(builder, scaffold, insn.operands[0]);
+    var [rhsRepr, rhsValue] = check buildReprValue(builder, scaffold, insn.operands[1]);
+    bir:Register result = insn.result;
+
+    match [lhsRepr.base, rhsRepr.base] {
+        [BASE_REPR_TAGGED, BASE_REPR_INT] => {
+            buildCompareTaggedInt(builder, scaffold, buildIntCompareOp(insn.op), lhsValue, rhsValue, result);
+        }
+        [BASE_REPR_INT, BASE_REPR_TAGGED] => {
+            buildCompareTaggedInt(builder, scaffold, buildIntCompareOp(flipOrderOp(insn.op)), rhsValue, lhsValue, result);
+        }
+        [BASE_REPR_TAGGED, BASE_REPR_FLOAT] => {
+            buildCompareTaggedFloat(builder, scaffold, buildFloatCompareOp(insn.op), lhsValue, rhsValue, result);
+        }
+        [BASE_REPR_FLOAT, BASE_REPR_TAGGED] => {
+            buildCompareTaggedFloat(builder, scaffold, buildFloatCompareOp(flipOrderOp(insn.op)), rhsValue, lhsValue, result);
+        }
+        [BASE_REPR_TAGGED, BASE_REPR_BOOLEAN] => {
+            buildCompareTaggedBoolean(builder, scaffold, buildBooleanCompareOp(insn.op), lhsValue, rhsValue, result);
+        }
+        [BASE_REPR_BOOLEAN, BASE_REPR_TAGGED] => {
+            buildCompareTaggedBoolean(builder, scaffold, buildBooleanCompareOp(flipOrderOp(insn.op)), rhsValue, lhsValue, result);
+        }
+        [BASE_REPR_TAGGED, BASE_REPR_TAGGED] => {
+            if insn.orderType is bir:OptOrderType {
+                // Both operands can be nil;
+                buildCompareTagged(builder, scaffold, insn, lhsValue, rhsValue, result);
+            }
+            else {
+                // Both operand are (nil) strings
+                buildCompareString(builder, scaffold, buildIntCompareOp(insn.op), lhsValue, rhsValue, result);
+            }
+        }
+        [BASE_REPR_INT, BASE_REPR_INT] => {
+            buildCompareInt(builder, scaffold, buildIntCompareOp(insn.op), lhsValue, rhsValue, result);
+        }
+        [BASE_REPR_BOOLEAN, BASE_REPR_BOOLEAN] => {
+            buildCompareInt(builder, scaffold, buildBooleanCompareOp(insn.op), lhsValue, rhsValue, result);
+        }
+        [BASE_REPR_FLOAT, BASE_REPR_FLOAT] => {
+            buildCompareFloat(builder, scaffold, buildFloatCompareOp(insn.op), lhsValue, rhsValue, result);
+        }
+    }
+}
+
+function flipOrderOp(bir:OrderOp op) returns bir:OrderOp {
```

Comment:
Less code to use a map.

`final readonly & map<bir:OrderOp> flippedOrderOps = { ... }`

Reply:
Fixed